`arrow2` erased refcounted clones benchmarks

teh-cmc opened the pull request:
Accompanying benchmarks for #1746.

Linux 5950x:
```
$ taskset -c 7 cargo bench --all-features -- arrow2/erased

arrow2/erased_clone/primitive/rows=10000/instances=100/cell/arc_erased
                        time:   [345.70 µs 346.16 µs 346.81 µs]
                        thrpt:  [28.834 Melem/s 28.889 Melem/s 28.927 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/cell/wrapped_in_arc
                        time:   [66.321 µs 66.392 µs 66.472 µs]
                        thrpt:  [150.44 Melem/s 150.62 Melem/s 150.78 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/array
                        time:   [333.01 µs 333.49 µs 334.08 µs]
                        thrpt:  [29.933 Melem/s 29.986 Melem/s 30.029 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/array/downcast_first
                        time:   [131.87 µs 131.93 µs 132.01 µs]
                        thrpt:  [75.754 Melem/s 75.796 Melem/s 75.831 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/vec/full_copy
                        time:   [379.46 µs 379.93 µs 380.44 µs]
                        thrpt:  [26.285 Melem/s 26.320 Melem/s 26.353 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/vec/wrapped_in_arc
                        time:   [52.433 µs 52.515 µs 52.609 µs]
                        thrpt:  [190.08 Melem/s 190.42 Melem/s 190.72 Melem/s]

arrow2/erased_clone/struct/rows=10000/instances=100/cell/arc_erased
                        time:   [2.1752 ms 2.1793 ms 2.1842 ms]
                        thrpt:  [4.5783 Melem/s 4.5886 Melem/s 4.5973 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/cell/wrapped_in_arc
                        time:   [88.589 µs 88.690 µs 88.798 µs]
                        thrpt:  [112.62 Melem/s 112.75 Melem/s 112.88 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/array
                        time:   [2.1399 ms 2.1447 ms 2.1519 ms]
                        thrpt:  [4.6470 Melem/s 4.6626 Melem/s 4.6731 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/array/downcast_first
                        time:   [1.8545 ms 1.8562 ms 1.8584 ms]
                        thrpt:  [5.3811 Melem/s 5.3873 Melem/s 5.3922 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/vec/full_copy
                        time:   [720.77 µs 723.63 µs 727.00 µs]
                        thrpt:  [13.755 Melem/s 13.819 Melem/s 13.874 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/vec/wrapped_in_arc
                        time:   [51.734 µs 51.769 µs 51.814 µs]
                        thrpt:  [193.00 Melem/s 193.17 Melem/s 193.30 Melem/s]
```

Wumpf commented:
M1 Max results:

```
cargo bench --all-features -- arrow2/erased

arrow2/erased_clone/rows=10000/instances=100/cell/arc_erased
                        time:   [1.9419 ms 1.9533 ms 1.9736 ms]
                        thrpt:  [5.0668 Melem/s 5.1197 Melem/s 5.1495 Melem/s]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
arrow2/erased_clone/rows=10000/instances=100/cell/wrapped_in_arc
                        time:   [80.812 µs 81.257 µs 81.819 µs]
                        thrpt:  [122.22 Melem/s 123.07 Melem/s 123.74 Melem/s]
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
arrow2/erased_clone/rows=10000/instances=100/array/downcast_first
                        time:   [22.771 µs 22.997 µs 23.305 µs]
                        thrpt:  [429.10 Melem/s 434.83 Melem/s 439.16 Melem/s]
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high severe
arrow2/erased_clone/rows=10000/instances=100/vec/full_copy
                        time:   [625.85 µs 627.36 µs 629.31 µs]
                        thrpt:  [15.891 Melem/s 15.940 Melem/s 15.978 Melem/s]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe
arrow2/erased_clone/rows=10000/instances=100/vec/wrapped_in_arc
                        time:   [53.257 µs 53.294 µs 53.338 µs]
                        thrpt:  [187.48 Melem/s 187.64 Melem/s 187.77 Melem/s]
Found 11 outliers among 100 measurements (11.00%)
  9 (9.00%) high mild
  2 (2.00%) high severe
  ```

Wumpf commented on crates/re_arrow_store/benches/arrow2.rs:
```diff
@@ -0,0 +1,148 @@
+
+            group.bench_function("array/downcast_first", |b| {
+                b.iter(|| {
+                    let arrays = arrays.clone();
```
This test doesn't make sense I believe. This is an array of references (Rust analyzer confirms my reading). I don't think copying `Vec<&T>` is what you had in mind for comparison? That would ofc massively outpace everything else.

teh-cmc replied:
Oh yeah nice catch, I forgot a `.clone()` up there

Wumpf commented on crates/re_arrow_store/benches/arrow2.rs:
```diff
@@ -0,0 +1,148 @@
+            group.bench_function("cell/arc_erased", |b| {
+                b.iter(|| {
+                    let cells = cells.clone();
+                    assert_eq!(
+                        total_instances,
+                        cells.iter().map(|cell| cell.num_instances()).sum::<u32>()
+                    );
```
This got me, so I started and left a tiny bit of research.

* each element in the vec is larger: A `DataCell`  has 40bytes (according to RA), Arcs are just 8 bytes so iterating alone is expected to be slower on `DataCell` (and `downcast_first` iterates & copies references :P)
* in order to clone a `DataCell`...
   * we clone both fields, i.e. the `ComponentName` and the  `Box<dyn arrow2::array:Array>`
       * cloning the `ComponentName` is cheap, it's just memcpying 2x 8 bytes, but that alone touches as much as cloning an Arc
       * cloning the `Box`
           * means we do another heap allocation
           * and then call clone on the `dyn arrow2::array:Array`, this **now** goes to `DynClone` as you mentioned on Slack. But I'd argue at this point we already had massive overhead compared to "copy 8 bytes and do an `atomicAdd(Acquire)` on another 8 bytes"
               * from here on it gets complicated, need more time for this... you take over ;D

Wumpf reviewed (changes requested):
`downcast_first` might not be what you intended (?), see comment

teh-cmc commented:
Alright, updated the benchmarks to cover most of your comments.

---

First, regarding `DataCell` containing an extra `ComponentName` making the comparison somewhat biased: you're right, I should have measured without the extra field (I didn't bother cause I didn't think it would change much..).

I've added a benchmark that measures cloning _only_ the erased array contained within the cell, rather than the entire cell:
```
# Cloning the whole DataCell
arrow2/erased_clone/struct/rows=10000/instances=100/cell/arc_erased
                        time:   [2.1752 ms 2.1793 ms 2.1842 ms]
                        thrpt:  [4.5783 Melem/s 4.5886 Melem/s 4.5973 Melem/s]

# Cloning just the erased array within the cell
arrow2/erased_clone/struct/rows=10000/instances=100/array
                        time:   [2.1399 ms 2.1447 ms 2.1519 ms]
                        thrpt:  [4.6470 Melem/s 4.6626 Melem/s 4.6731 Melem/s]
```

Behaves as expected: a very slight improvement.

---

Second, the benchmarks now cover both primitive arrays as well as compound arrays.

This is important because primitive arrays map 1-to-1 to erased arrays when you erase them: i.e. erasing a `PrimitiveArray<u64>` yields one and only on `Box<dyn Array>`.  
Compound arrays on the other hand are themselves collections of erased arrays, i.e. a `StructArray` is literally defined as:
```rust
#[derive(Clone)]
pub struct StructArray {
    data_type: DataType,
    values: Vec<Box<dyn Array>>,
    validity: Option<Bitmap>,
}
```
I.e. a `StructArray` would go through the `dyn-clone` machinery a bunch of times whether it's erased or not.

This is reflected in the `downcast_first` benchmark (which has been fixed, thanks for catching that :no_mouth:):
```
# Primitive data
arrow2/erased_clone/primitive/rows=10000/instances=100/array
                        time:   [333.01 µs 333.49 µs 334.08 µs]
                        thrpt:  [29.933 Melem/s 29.986 Melem/s 30.029 Melem/s]
arrow2/erased_clone/primitive/rows=10000/instances=100/array/downcast_first
                        time:   [131.87 µs 131.93 µs 132.01 µs]
                        thrpt:  [75.754 Melem/s 75.796 Melem/s 75.831 Melem/s]

# Compound data
arrow2/erased_clone/struct/rows=10000/instances=100/array
                        time:   [2.1399 ms 2.1447 ms 2.1519 ms]
                        thrpt:  [4.6470 Melem/s 4.6626 Melem/s 4.6731 Melem/s]
arrow2/erased_clone/struct/rows=10000/instances=100/array/downcast_first
                        time:   [1.8545 ms 1.8562 ms 1.8584 ms]
                        thrpt:  [5.3811 Melem/s 5.3873 Melem/s 5.3922 Melem/s]
```

As can be seen here, manually downcasting a primitive array has a massive impact on the cost of cloning it; not so much for a compound array, which makes sense.

---

All in all, my conclusion is still that the cost of all this machinery is unmanageable.

Whether the slowness comes from the extra allocations, the boxes, the `dyn-clone` machinery, the virtual calls, or a mix of all of them, I don't think it really matters since the solution is the same: just wrap the erased array in an `Arc` directly within the `DataCell` itself, and only pass `DataCell`s around. Don't work directly with erased arrow arrays if you can help it. Ever.

teh-cmc commented:
> Whether the slowness comes from the extra allocations, the boxes, the dyn-clone machinery, the virtual calls, or a mix of all of them, I don't think it really matters since the solution is the same: just wrap the erased array in an Arc directly within the DataCell itself, and only pass DataCells around. Don't work directly with erased arrow arrays if you can help it. Ever.

#1743 pretty much demonstrates that the virtual calls are what's wrecking us here and, due to the way `arrow2` is structured, the number of virtual calls actually **scales linearly** with the number of fields in a type.
I.e. cloning or estimating the size of a `Point3D` will be much slower than doing the same thing on an `InstanceKey`.

The solution is the same: get these operations out of the fast path, cache the results in `DataCell`s, use `DataCell`s everywhere.

Wumpf commented on crates/re_arrow_store/benches/arrow2.rs:
```diff
@@ -0,0 +1,229 @@
+// ---
+
+fn estimated_size_bytes(c: &mut Criterion) {
+    let kind = ["primitive", "struct"];
```
y'know an enum doesn't hurt even if its throw away, would take as much space as all these unrechables :P

teh-cmc replied:
I was tired and desperate lol; yeah let's make it into an enum

Wumpf commented on crates/re_arrow_store/benches/arrow2.rs:
```diff
@@ -0,0 +1,229 @@
+                let total_instances = arrays.iter().map(|array| array.len() as u32).sum::<u32>();
+                assert_eq!(total_instances, (NUM_ROWS * NUM_INSTANCES) as u32);
+
+                group.bench_function("array", |b| {
```
that's really `array/full_copy`, no?

teh-cmc replied:
no full copy here, this is the usual arrow path which ends at `arrow2::Buffer::clone()` which is an `Arc`

Wumpf reviewed (approved):
great improvements & writeup

updated the m1 numbers
